src/data/feature_index.js

```diff
@@ -125,79 +114,105 @@ class FeatureIndex {
 
         const matching = this.grid.query(minX - queryPadding, minY - queryPadding, maxX + queryPadding, maxY + queryPadding);
         matching.sort(topDownFeatureComparator);
-        this.filterMatching(result, matching, this.featureIndexArray, queryGeometry, filter, params.layers, styleLayers, pixelsToTileUnits, args.posMatrix, args.transform);
-
-        const matchingSymbols = args.collisionIndex ?
-            args.collisionIndex.queryRenderedSymbols(queryGeometry, this.tileID, args.tileSize / EXTENT, args.collisionBoxArray, args.sourceID, args.bucketInstanceIds) :
-            [];
-        matchingSymbols.sort();
-        this.filterMatching(result, matchingSymbols, args.collisionBoxArray, queryGeometry, filter, params.layers, styleLayers, pixelsToTileUnits, args.posMatrix, args.transform);
+        const result = {};
+        for (const index of matching) {
+            const match = this.featureIndexArray.get(index);
+            let geometry = null;
+            this.generateGeoJSONFeature(
+                result,
+                match.bucketIndex,
+                match.sourceLayerIndex,
+                match.featureIndex,
+                filter,
+                params.layers,
+                styleLayers,
+                (feature: VectorTileFeature, styleLayer: StyleLayer) => {
+                    if (!geometry) {
+                        geometry = loadGeometry(feature);
+                    }
+                    return styleLayer.queryIntersectsFeature(queryGeometry, feature, geometry, this.z, args.transform, pixelsToTileUnits, args.posMatrix);
+                }
+            );
+        }
 
         return result;
     }
 
-    filterMatching(
+    generateGeoJSONFeature(
         result: {[string]: Array<{ featureIndex: number, feature: GeoJSONFeature }>},
-        matching: Array<any>,
-        array: FeatureIndexArray | CollisionBoxArray,
-        queryGeometry: Array<Array<Point>>,
+        bucketIndex: number,
+        sourceLayerIndex: number,
+        featureIndex: number,
         filter: FeatureFilter,
         filterLayerIDs: Array<string>,
         styleLayers: {[string]: StyleLayer},
-        pixelsToTileUnits: number,
-        posMatrix: Float32Array,
-        transform: Transform
-    ) {
-        let previousIndex;
-        for (let k = 0; k < matching.length; k++) {
-            const index = matching[k];
+        intersectionTest?: (feature: VectorTileFeature, styleLayer: StyleLayer) => boolean) {
 
-            // don't check the same feature more than once
```

Comment:
🤦‍♂️ This turned out to be the cause of the performance regression -- we definitely still need to filter out duplicate entries. I misread `FeatureIndex#insert`, which can actually make _very many_ entries in the grid for a single feature. I spent half a day messing around trying to reliably profile QueryBox before I finally realized this...